cpp/src/join/cross_join.cu

```diff
@@ -0,0 +1,77 @@
+#include <cudf/column/column.hpp>
+#include <cudf/copying.hpp>
+#include <cudf/detail/nvtx/ranges.hpp>
+#include <cudf/detail/repeat.hpp>
+#include <cudf/filling.hpp>
+#include <cudf/join.hpp>
+#include <cudf/reshape.hpp>
+#include <cudf/table/table.hpp>
+#include <cudf/table/table_view.hpp>
+#include <cudf/types.hpp>
+#include <cudf/utilities/error.hpp>
+
+namespace cudf {
+namespace detail {
+/**
+ * @brief  Performs a cross join on two tables (left, right)
+ *
+ * The cross join returns the cartesian product of rows from each table.
+ *
+ * The approach is to repeat the left table by the number of rows in the right table
+ * and tile the right table by the number of rows in the left table.
+ *
+ * @throws cudf::logic_error if number of columns in either `left` or `right` table is 0
+ *
+ * @param[in] left                  The left table
+ * @param[in] right                 The right table
+ * @param[in] mr                    Device memory resource to use for device memory allocation
+ * @param[in] stream                Cuda stream
+ *
+ * @returns                         Result of cross joining `left` and `right` tables
+ */
+std::unique_ptr<cudf::table> cross_join(
+  cudf::table_view const& left,
+  cudf::table_view const& right,
+  rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource(),
+  cudaStream_t stream                 = 0)
```

Comment:
Please leave the default parameter for `mr`. One reason for the rearrangement was so that we don't have to write `rmm::mr::get_default_resource()` in every call. #5119

Reply:
I will swap `stream` and `mr` in the new `detail` APIs introduced in this PR. This will need to be refactored in https://github.com/rapidsai/cudf/pull/5135.